### ISSUE TITLE
List of non-python 3 packages

### ISSUE THREAD

**Issue by ocefpaf** [ioos/conda-recipes#429]
- [X] `ulmo`
- [X] `suds` (suds-jurko)
- [X] `pyke`
- [X] `unit_conversion`
- [X] `numdifftools` (fixed in #574)
- [X] `pyproj`
- [ ] `pyoos`
- [ ] `pycsw` (WIP)
- [ ] `iris` (WIP)
- [ ] `utilities` (`iris`)
- [ ] `tardis` (`iris`)
- [ ] `pywafo`
- [ ] `octant`
- [ ] `pastescript` (will not fix)
- [ ] `pygdp` (WIP)
- [ ] `optcomplete` (will not fix)
- [ ] `petulant-bear` (will not fix)
- [ ] `pyhum`
- [ ] `pydap` (will not fix)
- [ ] `pyfvcom` (will not fix)
- [ ] `compliance-checker`
- [ ] `cc-plugin-glider` (`compliance-checker`)
- [ ] `pyseidon`
- [ ] `wicken`
- [ ] `mocsy` (compiler issue #567)

**Comment by kwilcox** [comment]
It's super impressive that this list is so short!  Awesome stuff @ocefpaf

**Comment by ocefpaf** [comment]
I might be missing something, but those are known failure.  If you ever hit a wall let me know,

PS: pyke has a Python3 version, but we are having issues with it now.  Also, suds has a substitute for Python3 (`suds-jurko`), but I am planning to deprecated that package.  I guess we no longer use it.

**Comment by tomkralidis** [comment]
We have an an open [issue](https://github.com/geopython/pycsw/issues/328) in pycsw for Python 3 support.  @QuLogic has done excellent work in helping bump [OWSLib](https://github.com/geopython/OWSLib) and [geolinks](https://github.com/geopython/geolinks) to Python 3 support, but I'm not sure about the rest of pycsw dependencies (SQLAlchemy, pyproj).  Any pycsw Python 3 help would be very much appreciated.

**Comment by ocefpaf** [comment]
I `pyproj` is already Python3 (and available in the IOOS channel :wink: )

I did not test `SQLAlchemy`, but the [classifiers at PyPI](https://pypi.python.org/pypi/SQLAlchemy) say Python3 is available.

I am not sure the amount of work that is needed in pycsw, but I will add to my list for a long rainy weekend.

**Comment by QuLogic** [comment]
For some reason [pyproj is listed as a blocker](https://caniusepython3.com/check/b39a31bd-4fa1-4021-b740-08c019b99757); maybe they need to update their classifiers on PyPI. OWSLib is a blocker only because it's limited to <0.9.0.

**Comment by tomkralidis** [comment]
Looks like @ocefpaf is working on it and getting there: https://github.com/ocefpaf/pycsw/tree/py3k

**Comment by ocefpaf** [comment]
It is raining here...

**Comment by rsignell-usgs** [comment]
I am contributing by not harrassing @ocefpaf this afternoon.  :smile_cat:

**Comment by ocefpaf** [comment]
> I am contributing by not harrassing @ocefpaf this afternoon.  :smile_cat:

:stuck_out_tongue_winking_eye:

**Comment by emiliom** [comment]
That's great, @ocefpaf ! Regarding suds:
> Also, suds has a substitute for Python3 (suds-jurko), but I am planning to deprecated that package. I guess we no longer use it.

suds is an [ulmo](https://github.com/ulmo-dev/ulmo) dependency, for SOAP access, for CUAHSI WaterOneFlow/WaterML 1 access. If the intent is to keep ulmo on the ioos channel, suds would have to stay. I'm aware that suds has been abandoned, and I've seen suds-jurko but never tried it; don't know how much effort it'd be to replace suds with suds-jurko in ulmo. But @dharhas is [currently overhauling ulmo](https://github.com/ulmo-dev/ulmo/issues/109), so this wouldn't be a time to invest in the current codebase. After that refactoring, my understanding is that the suds dependency would likely become an optional plugin.

Alternatively, @dharhas also has an [ulmo conda recipe and binstar channel](https://binstar.org/ulmo), but I don't know how actively maintained it is. FYI, I use ulmo regularly, or whether relying on an outside channel may create inconsistencies with stuff from the ioos channel.

**Comment by ocefpaf** [comment]
> If the intent is to keep ulmo on the ioos channel, suds would have to stay.

Our deprecation process remove unsupported recipes from the repository but the binaries do stay in the channel.  Since suds is unsupported it makes sense to deprecate it.  And all the packages that depends on suds will have to do with the last available version.

The reason for deprecation is to lessen the burden on our numerous maintainers :wink:

PS: I am watching @dharhas' ulmo refactoring close and I will keep our recipe up-to-date.

**Comment by emiliom** [comment]
Thanks for the explanation. Just goes to show my ignorance on binstar channel & recipe repo intricacies.

Glad you're watching the ulmo refactoring! I'm starting to suspect that days are longer than 24 hours out there in the Nordeste :) Or you have a clone.

**Comment by ocefpaf** [comment]
> than 24 hours out there in the Nordeste :) Or you have a clone.

Nope.  Just lots of rain due to a strong ENSO.  We are not getting many chances to get to the beach this year.  Well.. At least we will get more coding done :wink:

**Comment by dharhas** [comment]
@ocefpaf @emiliom 

so for kicks and giggles a few days ago, I ran python futurize on the current version of ulmo and tried running the tests on python 3. About 70% of the tests passed and quite a few of the rest seemed to be a unicode vs string issue. So that gives me more confidence about making ulmo python 3 compatible. I'm hoping that suds-jurko is a drop in replacement for suds. If it isn't I'm not sure what alternatives we have. 

If possible I'd like to make the IOOS channel the primary place to get ulmo and drop my binstar/anaconda channel. I don't really maintain the ulmo channel very well and don't have an automated build system in place.

In another note, I'm considering keeping ulmo as a single package rather than splitting it up into optional packages. Basically multiple packages makes for more overhead in terms on managing releases/running tests etc and since I'm pretty much the only one maintaining ulmo I'm not sure it makes sense to split it up.

Bottom line. In the near (for some value of near) future ulmo should be py3 compatible. I need it to be py3 compatible for another project.

- dharhas

**Comment by ocefpaf** [comment]
>  About 70% of the tests passed and quite a few of the rest seemed to be a unicode vs string issue.

Awesome!

> If possible I'd like to make the IOOS channel the primary place to get ulmo.

It definitely is.  Just ping us when you have a newer version of ulmo and we will build it :wink:

**Comment by emiliom** [comment]
Re: ulmo -- Great on all counts ("consolidating" to IOOS channel, py3 future, etc).

Great also to see @ocefpaf has [just added a recipe for suds-jurko](https://github.com/ioos/conda-recipes/pull/431)! Like @dharhas said, hopefully it'll be a drop-in replacement for suds. Dharhas, I can volunteer to help test it, sometime after this week.

**Comment by ocefpaf** [comment]
Pinging @lukecampbell :exclamation:
(See the last item on the list.)

**Comment by rsignell-usgs** [comment]
But it is kind of amusing that the `compliance-checker` is not Python 3-compliant, no? :stuck_out_tongue_closed_eyes:

**Comment by lukecampbell** [comment]
I'm knee deep in some major changes to compliance checker, I'll see if I can't bring it up to compliance soon.

**Comment by dharhas** [comment]
@emiliom @ocefpaf @rsignell-usgs 

So after much unicode/string/bytes pain (I still don't think I completely get how all of that works), I have a py2.7/3.4 compatible version of ulmo that is passing tests on travis. Planning to merge it in as ulmo 0.8, checkout https://github.com/ulmo-dev/ulmo/pull/112

if any of you are using ulmo in scripts please try them with py2/3. Tests are passing but I don't have complete coverage so ymmv.

**Comment by ocefpaf** [comment]
> So after much unicode/string/bytes pain

So you played everybody favorite game `encode`/`decode` whack-a-mole :wink: 

> (I still don't think I completely get how all of that works)

Me neither.  I just hope to do this kind of thing only once.

> Planning to merge it in as ulmo 0.8

As soon as you merge I will update the package here.  Thanks @dharhas!

**Comment by dharhas** [comment]
yup. plus some BytesIO/StringIO whack-a-mole for good measure.

**Comment by dharhas** [comment]
@ocefpaf 

ulmo 0.8.0 is tagged v0.8.0 on github and uploaded to pypi. I have CI setup with AppVeyor and Travis and everything looks good now. Look at the py2/3_conda_environment.yml files to see the packages ulmo needs. There is a hard crash for pytables > 3.1.1 on windows/py3.4

**Comment by ocefpaf** [comment]
Almost there.  We need https://github.com/ulmo-dev/ulmo/pull/113.

Quick questions:

1) Is mock a dependency or a test dependency?  It is not listed in the `tests_require` and the comment on `setup.py` makes me think it is a required dependency.

2) `lxml` is listed in `setup.py`, but absent in the `py2/3_conda_environment.yml`.  I guess it is still a dependency, right?

**Comment by dharhas** [comment]
Ok I've merged the pull request. Before I tag a 0.8.1 release I thought I'd check a few things.

1) After checking, mock is used in two places. a) For tests and b) To enable readthedocs to build the docs without having to install pytables (pytables is an optional dependency). Without pytables the usgs.nwis.hdf5 stuff doesn't work and all hdf5 related tests fail.

2) lxml is not used directly by ulmo but is used by beautifulsoup4. When you conda install bs4 you get lxml automatically but I think in general it is a optional dependency of bs4 so setup.py explicitly lists it.

So should I explicitly add lxml to the conda_environment files?

**Comment by ocefpaf** [comment]
> Without pytables the usgs.nwis.hdf5 stuff doesn't work and all hdf5 related tests fail.

OK.  I will add it as dependency.

> So should I explicitly add lxml to the conda_environment files?

I think that it is best. You never know if you will not `lxml` or not when installing bs4.  And I see that ulmo imports `lxml` in a few places ([here](https://github.com/ulmo-dev/ulmo/blob/875bdbc61d7223af8fd47505cc139ccfed8e90f0/ulmo/waterml/common.py#L3) and [here](https://github.com/ulmo-dev/ulmo/blob/875bdbc61d7223af8fd47505cc139ccfed8e90f0/ulmo/util/misc.py#L18)).

**Comment by dharhas** [comment]
Ok 0.8.1 is on pypi now and ulmo is pip installable (and I worked out how to use the pypitest server before making the actual release, so I don't use up version numbers on broken releases)

**Comment by ocefpaf** [comment]
PR #434 is using PyPI.  Thanks @dharhas!

**Comment by QuLogic** [comment]
For Python 3.3+, you could just import `unittest.mock` and leave out the `mock` dependency. The package is just a backport, and up until the release of 3.5, probably wouldn't be much difference.

**Comment by ocefpaf** [comment]
Thanks @QuLogic!  I will look into that.

**Comment by ocefpaf** [comment]
This list is outdated and most of these packages either already have Python 3 support or is abandoned and never will.
